dev/benchmarking/ddl/1_tables.sql

```diff
@@ -0,0 +1,291 @@
+  IS 'The benchmark language version, e.g. ''2.7'' for Python 2.7';
+COMMENT ON COLUMN public.language_version.language_version
+  IS 'The version number used in the benchmark environment (e.g. ''2.7'').';
+
+-- BENCHMARK_TYPE
+CREATE TABLE IF NOT EXISTS public.benchmark_type
+(
+  id SERIAL PRIMARY KEY
+  , benchmark_type citext NOT NULL UNIQUE
+);
+COMMENT ON TABLE public.benchmark_type
+  IS 'The type of benchmark. For example "time", "mem", "peakmem", "track"';
+COMMENT ON COLUMN public.benchmark_type.benchmark_type
+  IS 'The type of units, so "time" for sec/ns/ms or "mem" for kB, MB, GB.';
+
+-- MACHINE
+CREATE TABLE IF NOT EXISTS public.machine
+(
+  id SERIAL PRIMARY KEY
+  , machine_name citext NOT NULL UNIQUE
+  , memory_bytes bigint NOT NULL
+  , cpu_actual_frequency_Hz bigint NOT NULL
+  , cpu_id integer NOT NULL
+  , gpu_id integer NOT NULL
+  , os_id integer NOT NULL
+  , FOREIGN KEY (cpu_id) REFERENCES public.cpu
+  , FOREIGN KEY (gpu_id) REFERENCES public.gpu
+  , FOREIGN KEY (os_id) REFERENCES public.os
+);
+COMMENT ON TABLE public.machine
+  IS 'Unique identifiers for a machine.';
+COMMENT ON COLUMN public.machine.machine_name
+  IS 'A unique machine name of your choice.';
+COMMENT ON COLUMN public.machine.memory_bytes
+  IS 'The output of `sysctl -n hw.memsize`.';
+COMMENT ON COLUMN public.machine.cpu_actual_frequency_Hz
+  IS 'The output of `sysctl -n hw.cpufrequency`.';
+
+-- UNIT
+CREATE TABLE IF NOT EXISTS public.unit
+(
+  id SERIAL PRIMARY KEY
+  , unit_name citext NOT NULL UNIQUE
+  , benchmark_type_id integer NOT NULL
+  , FOREIGN KEY (benchmark_type_id) REFERENCES public.benchmark_type(id)
+);
+COMMENT ON TABLE public.unit IS 'The actual units for a reported benchmark.';
+COMMENT ON COLUMN public.unit.unit_name
+  IS 'For example: ms, ns, microseconds.';
+
+-- ENVIRONMENT
+CREATE TABLE IF NOT EXISTS public.environment
+(
+  id SERIAL
+  , benchmark_language_id integer NOT NULL
+  , language_version_id integer NOT NULL
+  , language_dependencies_id integer NOT NULL
+  , PRIMARY KEY
+      (id, benchmark_language_id, language_version_id)
+  , FOREIGN KEY
+      (benchmark_language_id)
+      REFERENCES public.benchmark_language
+  , FOREIGN KEY
+      (language_version_id, benchmark_language_id)
+      REFERENCES public.language_version(id, benchmark_language_id)
+  , FOREIGN KEY
+      (language_dependencies_id, benchmark_language_id)
+      REFERENCES public.language_dependencies(id, benchmark_language_id)
+);
+COMMENT ON TABLE public.environment
+  IS 'Identifies a build environment for a specific suite of benchmarks.';
+
+-- BENCHMARK
+CREATE TABLE IF NOT EXISTS public.benchmark
+(
+  id SERIAL
+  , benchmark_language_id integer NOT NULL
+  , benchmark_name citext NOT NULL
+  , parameter_names text[]
+  , lessisbetter boolean NOT NULL
+  , benchmark_description text NOT NULL
+  , benchmark_version citext NOT NULL
+  , unit_id integer NOT NULL
+  , PRIMARY KEY (id, benchmark_language_id)
```

Comment:
why the compound key, when there is already a synthetic key?

Reply:
This is again related to pulling the `benchmark_language` through all of its dependent tables. Agree that it is complex, hard to read, and possibly hard to maintain downstream. Advice welcome for how to accomplish this in a more elegant way.

Reply:
one interesting question is if a benchmark should should bind the language  or if only a benchmark_result should do it.  I think the more common case would be the benchmark would have the language, but I can also make the counter argument for example, how long does it take to create arrow arrays from a parquet file, each language would be slightly different but still might be interesting to compare.

Reply:
I honestly am at a loss here. I have experience with databases (though currently feeling a bit humbled) but have never done benchmarking before and am completely reliant on the feedback of others to make the tables not suck.

I'm attempting to model the data structures after the ASV library since there are ASV benchmarks already in [arrow/python/benchmarks](https://github.com/apache/arrow/tree/master/python/benchmarks).

It is very, very useful to have your comments. Thank you for taking so much time to work through this with me. I kept the original compound key structure with the languages in the newest push just to get something out; but am happy to modify.